eth/executionclient/execution_client.go

```diff
@@ -0,0 +1,391 @@
+
+	"github.com/ethereum/go-ethereum"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"go.uber.org/zap"
+
+	"github.com/bloxapp/ssv/eth/contract"
+	"github.com/bloxapp/ssv/logging/fields"
+	"github.com/bloxapp/ssv/utils/tasks"
+)
+
+var (
+	ErrClosed       = fmt.Errorf("closed")
+	ErrNotConnected = fmt.Errorf("not connected")
+	ErrBadInput     = fmt.Errorf("bad input")
+)
+
+// ExecutionClient represents a client for interacting with Ethereum execution client.
+type ExecutionClient struct {
+	// mandatory
+	nodeAddr        string
+	contractAddress ethcommon.Address
+
+	// optional
+	logger                      *zap.Logger
+	metrics                     metrics
+	finalizationOffset          uint64
+	connectionTimeout           time.Duration
+	reconnectionInitialInterval time.Duration
+	reconnectionMaxInterval     time.Duration
+	logBatchSize                uint64
+
+	// variables
+	client    atomic.Pointer[ethclient.Client]
+	connectMu sync.Mutex
+	closed    chan struct{}
+}
+
+// New creates a new instance of ExecutionClient.
+func New(nodeAddr string, contractAddr ethcommon.Address, opts ...Option) *ExecutionClient {
+	client := &ExecutionClient{
+		nodeAddr:                    nodeAddr,
+		contractAddress:             contractAddr,
+		logger:                      zap.NewNop(),
+		metrics:                     nopMetrics{},
+		finalizationOffset:          DefaultFinalizationOffset,
+		connectionTimeout:           DefaultConnectionTimeout,
+		reconnectionInitialInterval: DefaultReconnectionInitialInterval,
+		reconnectionMaxInterval:     DefaultReconnectionMaxInterval,
+		logBatchSize:                DefaultHistoricalLogsBatchSize, // TODO Make batch of logs adaptive depending on "websocket: read limit"
+		closed:                      make(chan struct{}),
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	return client
+}
+
+// Connect connects to Ethereum execution client.
+func (ec *ExecutionClient) Connect(ctx context.Context) {
+	ec.connectMu.Lock()
+	defer ec.connectMu.Unlock()
+
+	if ec.client.Load() != nil {
+		ec.reconnect(ctx)
+		return
+	}
+
+	if err := ec.connect(ctx); err != nil {
+		ec.reconnect(ctx)
+	}
+}
+
+// Close shuts down ExecutionClient.
+func (ec *ExecutionClient) Close() error {
+	close(ec.closed)
+
+	if client := ec.client.Load(); client != nil {
+		client.Close()
+	}
+
+	return nil
+}
+
+// FetchHistoricalLogs retrieves historical logs emitted by the contract starting from fromBlock.
+func (ec *ExecutionClient) FetchHistoricalLogs(ctx context.Context, fromBlock uint64) (logCh <-chan ethtypes.Log, fetchErrCh <-chan error, err error) {
+	client := ec.client.Load()
+	if client == nil {
+		return nil, nil, ErrNotConnected
+	}
+
+	currentBlock, err := client.BlockNumber(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("get current block: %w", err)
+	}
+
+	lastBlock := currentBlock - ec.finalizationOffset
```

Comment:
so finalizationOffset is constant? 
I think we initially thought to do it with reading the finalized checkpoint from consensus layer and go by it. @moshe-blox you think constant will be enough?

Reply:
As I understood, yes. If it shouldn't be, let's discuss the requirements. Added a TODO for this

Reply:
> so finalizationOffset is constant? I think we initially thought to do it with reading the finalized checkpoint from consensus layer and go by it. @moshe-blox you think constant will be enough?

For prater, yes. For mainnet, I think it's too risky because if there is a reorg, everything will break (unless no validators were added in the reorged blocks).

Like we said, it's possible to not wait so long for finalization and instead handle reorgs, but since it's too complex, we preferred to just wait for blocks to be finalized before reacting on them.

`finalizationOffset` is a good idea for us to test on stage without having to wait 2 epochs, but we should definitely wait for actual finalization in a permissionless mainnet environment.

@lior-blox do you think we should do this in a separate PR?